Go bindings

rafaelsierra opened the pull request:
This PR builds on top of #292

Commit:
go dir

Commit:
examples added

Commit:
config file for go

Commit:
test file

srinivasyadav18 committed:
"added go"

Commit:
added go

Commit:
added go

Commit:
test added

srinivasyadav18 committed:
example for indexing

srinivasyadav18 committed:
Update Makefile.am

srinivasyadav18 committed:
Update Makefile.am

srinivasyadav18 committed:
Update Makefile.am

srinivasyadav18 committed:
Update Makefile.am

Commit:
docs and gitignore

Commit:
ex and mk

Commit:
changed git ignore

Commit:
go bindings

srinivasyadav18 committed:
Dockerfile for xapian

srinivasyadav18 committed:
Update Makefile.am

srinivasyadav18 committed:
Update index.rst

srinivasyadav18 committed:
Update Makefile.am

srinivasyadav18 committed:
for tests

srinivasyadav18 committed:
Update Makefile.am

Commit:
tabs

Commit:
tabs and spaces

srinivasyadav18 committed:
Update Makefile.am

srinivasyadav18 committed:
Update Makefile.am

Commit:
update for os x

srinivasyadav18 committed:
Update Makefile.am

srinivasyadav18 committed:
white spaces

Commit:
removed extra white spaces

Commit:
check 3

srinivasyadav18 committed:
Update Makefile.am

srinivasyadav18 committed:
Update Makefile.am

srinivasyadav18 committed:
Update Makefile.am

srinivasyadav18 committed:
Update Makefile.am

srinivasyadav18 committed:
Update Makefile.am

srinivasyadav18 committed:
Update Makefile.am

srinivasyadav18 committed:
Update Makefile.am

srinivasyadav18 committed:
Update Makefile.am

srinivasyadav18 committed:
Update Makefile.am

srinivasyadav18 committed:
Update Makefile.am

srinivasyadav18 committed:
Update Makefile.am

srinivasyadav18 committed:
Update Makefile.am

srinivasyadav18 committed:
Update Makefile.am

srinivasyadav18 committed:
Update Makefile.am

srinivasyadav18 committed:
Update Makefile.am

srinivasyadav18 committed:
Update Makefile.am

Commit:
check

srinivasyadav18 committed:
Update Makefile.am

Commit:
test1

Commit:
test2

Commit:
rpath

srinivasyadav18 committed:
Update Makefile.am

srinivasyadav18 committed:
test

srinivasyadav18 committed:
Xapian go bindings

testing bindings without installing works :-)

Commit:
go path

Commit:
clang stdlib for cgo

Commit:
clang lib

Commit:
fix bug libc++ only

Commit:
fix bug clang -lc++

Commit:
fix bug clang -lc++

Commit:
improving from mentor's review

Commit:
use cxx_flags from configure

Commit:
cxxflags

Commit:
awk in one go

srinivasyadav18 committed:
remove hard coded ld flags

srinivasyadav18 committed:
Add LDFLAGS

Commit:
removing hard-coded things

Commit:
Remove hard-coding ldflags

Commit:
clang ld flag

Commit:
clang ld flag

srinivasyadav18 committed:
use clang lib for go

Commit:
fix bug clang lib

Commit:
fix bug clang-go

Commit:
la path

Commit:
fix

Commit:
path for lib

srinivasyadav18 committed:
clang path

srinivasyadav18 committed:
Update configure.ac

srinivasyadav18 committed:
Update configure.ac

srinivasyadav18 committed:
Update configure.ac

Commit:
path for clang

Commit:
lc

srinivasyadav18 committed:
"update Xapian-headers.i for Go"

Commit:
fixing from patch

Commit:
Bindings for linux,windows(MSYS),max

Commit:
Bindings with cross-platform setup.go

Commit:
appveyor config for bindings

Commit:
appveyor config for bindings

Commit:
appveyor config for bindings

Commit:
appveyor config for bindings

Commit:
testing again with old appveyor yml

Commit:
appveyor config for go-bindings

Commit:
mingw x86 build

Commit:
mingw x86 build

Commit:
mingw x86 build

Commit:
mingw x86 build

Commit:
checking x86

Commit:
fix appveyor build

Commit:
go bindings mingw64 x64

srinivasyadav18 committed:
clean up

srinivasyadav18 committed:
fix up spaces

srinivasyadav18 committed:
fix up spaces

srinivasyadav18 committed:
fix up spaces

srinivasyadav18 committed:
fix up spaces

rafaelsierra committed:
Updated build so that make and make install works

rafaelsierra committed:
Formatted code

rafaelsierra committed:
Beginning work to handle errors

rafaelsierra committed:
Dirty implementation for handling errors

rafaelsierra committed:
added back allow_failures

ojwb commented on xapian-bindings/go/xapian.go:
```diff
@@ -0,0 +1,53 @@
+package xapian
```
It doesn't seem ideal that this seems to require writing a wrapper for every wrapped function.  We use SWIG because it automates away all that drudgery.  We've had hand-coded Perl XS and Java JNI bindings in the past and they always ended up lagging behind the C++ API because of the extra work needed, whereas for SWIG-generated bindings in most cases a new method gets wrapped automatically, while new classes just need setting up according to what sort of classes they are (e.g. user-subclassable classes get special handling).

I'll take a look and see what we can do.

I think ideally SWIG itself would support a more Go-like way to wrap exceptions, but there doesn't seem to be much progress on that front.

ojwb commented on xapian-bindings/go/xapian.go:
```diff
@@ -0,0 +1,53 @@
+	return raw.NewTermGenerator()
+}
+
+func NewDatabase(args ...interface{}) (db raw.Database, err error) {
+	defer func() {
+		if r := recover(); r != nil {
+			if errMsg, ok := r.(string); ok {
+				for msgPrefix, errType := range errorMap {
+					if strings.HasPrefix(errMsg, msgPrefix) {
+						err = errType
+						return
+					}
+				}
+			}
+			panic(r)
+		}
+	}()
+
+	db = raw.NewDatabase(args...)
+	return
+}
```
BTW, I notice SWIG itself has moved away from using `crosscall2` for this in swig/swig@6ca5d5d722b07977cd2865337d4a9e6cf6d75d5f:

> swig -go: don't use crosscall2 for panicking
>    
> Instead rely only on documented and exported interfaces.

SWIG/Go doesn't seem to make use of `SWIG_RuntimeError`, etc but if you want to turn the C++ exception into a numeric code you can probably rethrow and catch it in C++ to dispatch by exception type.  Look at `SetPythonException()` in `xapian-bindings/python/except.i` (generated) for example.

ojwb commented on xapian-bindings/xapian-headers.i:
```diff
@@ -245,7 +245,7 @@ STANDARD_IGNORES(Xapian, Query)
 #if defined SWIGCSHARP || defined SWIGJAVA || defined SWIGPERL || \
-    defined SWIGPYTHON || defined SWIGRUBY
+    defined SWIGPYTHON || defined SWIGRUBY || defined SWIGGO
 // C#, Java, Perl, Python and Ruby wrap these "by hand" to give a nicer API
```
Comment no longer matches the code.

ojwb commented on xapian-core/INSTALL:
```diff
@@ -94,6 +94,26 @@ find the zlib headers and library.
+
+	make
+
+	make install
```
This would be useful to document (if the commands given actually work - I'm dubious how well tested it is, given the `--nocofirm` in the second pacman command seems like a typo) but isn't related to Go support so would be better as a separate patch.

ojwb commented on xapian-bindings/configure.ac:
```diff
@@ -1108,6 +1141,50 @@ if test no != "$with_lua" ; then
+  AC_MSG_RESULT([${USE_CLANG_LIB}])
+  if test -n "${USE_CLANG_LIB}" ; then
+    AC_MSG_RESULT([${USE_CLANG_LIB}])
+    AC_SUBST(USE_CLANG_GO_LIB,'-lc++')
```
Copying over a comment from the old PR which was never responded to:

> This also looks very dubious to me, and could misdetect if the stdlib is present elsewhere - e.g. GXX=/home/stdlib/bin/gcc, or misdetect the other way if CXXFLAGS=-stdlib=libc++ is used instead of CXX='clang -stdlib=libc++'.
>
> What is this actually trying to achieve?

I realise you may well not know, but if so it seems better to drop this and see what (if anything) breaks so we actually understand the need for it, and can perhaps address it better.

ojwb commented on .appveyor.yml:
```diff
@@ -69,16 +69,19 @@ matrix:
-          $env:CONFIGOPTS+=" --host=x86_64-w64-mingw32"
+          $env:CONFIGOPTS+=" --host=x86_64-w64-mingw32 --with-go"
+          $env:BINDINGS+=" xapian-bindings"
         }
```
In https://github.com/xapian/xapian/pull/292#pullrequestreview-462218704 I noted that the use of `$BINDINGS` was removed, apparently while resolving a merge conflict.

rafaelsierra committed:
Testing with go errors and exceptions

This commit also makes sure that the `xapian.go` generated on the source
tree contains the necessary CGO flags to compile

rafaelsierra committed:
Updated test to check for error type

rafaelsierra committed:
Removed Dockerfile

rafaelsierra commented on Dockerfile:
```diff
@@ -0,0 +1,21 @@
+
+#run these commands once you are in
+#source ~/.bashrc
+#cd xapian && git pull && ./bootstrap && ./configure && make && make install
```
agreed

rafaelsierra commented on xapian-bindings/go/go.i:
```diff
@@ -0,0 +1,66 @@
+%}
+
+%insert(go_wrapper) %{
+func NewDatabase(a ...interface{}) (db Database, err error) {
+	defer func() {
+		if r := recover(); r != nil {
+			if errMsg, ok := r.(string); ok {
+				err = errors.New(errMsg)
+				return
+			}
+			panic(r)
+		}
+	}()
+
+	db = NewWrappedDatabase(a...)
+	return
+}
```
I forgot to mention that I am, of course, using `Xapian::Database::Database` just as example, and that other functions that can also throw errors would follow the same pattern
